### DISCUSSION TITLE
Capture and show the response resulting from a call out to external server

### DISCUSSION THREAD

**Discussion by nvijayap** [vueform/vueform#74] [Questions] [answered]
Hi,

What's the canonical way to use `FormErrors/FormMessages` in `Vueform` (`to capture and show the response resulting from a call out to external server`) ?

This does not yield any errors/messages in browser (`NOTE`: `Equivalent call from command-line yields the errors/messages`) ...

```
<template>
  <Vueform endpoint="https://xxxxx.xx/xxxxx" method="post">
    <TextElement name="name" label="Name" rules="required|max:80" />
    <ButtonElement name="submit" submits> Submit </ButtonElement>
    <template #empty>
      <FormErrors v-if="hasMessages" />
    </template>
  </Vueform>
</template>
```

**Comment by adamberecz** [comment, marked as answer]
## Displaying error / success messages after submit

This would be one way to display error messages:

```vue
<script setup>
const handleResponse = (response, form$) => {
  form$.messageBag.clear() // clear custom messages before appending new ones

  if (response.data.success) { // or the way you check for success
    form$.messageBag.append('Success', 'message')
  } else {
    form$.messageBag.append('Error')
  }
}
</script>

<template>
  <Vueform endpoint="https://xxxxx.xx/xxxxx" method="post" @response="handleResponse">
    <TextElement name="name" label="Name" rules="required|max:80" />
    <ButtonElement name="submit" submits> Submit </ButtonElement>
  </Vueform>
</template>
```

You can read more custom messages here:
https://vueform.com/docs/validating-elements#custom-errors-and-messages

`<Vueform>` also have [`@success`](https://vueform.com/reference/vueform#event-success) and [`@error`](https://vueform.com/reference/vueform#event-success) events.

## Using `FormErrors` / `FormMessages` under elements

**Way 1:** If you want to show the errors under the form:
```vue
<template>
  <Vueform endpoint="https://xxxxx.xx/xxxxx" method="post" @response="handleResponse">
    <template #empty>
      <FormElements>
       <TextElement name="name" label="Name" rules="required|max:80" />
       <ButtonElement name="submit" submits> Submit </ButtonElement>
      </FormElements>
      <FormErrors v-if="hasMessages" />
    </template>
  </Vueform>
</template>
```
*(you were quite close)*

**Way 2:** You could create a custom version of Vueform component as an [SFC](https://vueform.com/docs/rendering-forms#using-single-file-component), based on the default [`<Vueform>` template](https://github.com/vueform/vueform/blob/main/themes/blank/templates/Vueform.vue) and reuse it over time:

```vue
<!-- CustomVueform.vue -->
<template>
  <form :class="classes.form" @submit.prevent="submit">
    <slot name="empty">
      <FormElements><slot/></FormElements>
      <component is="FormMessages" v-if="showMessages"/>
      <component is="FormErrors" v-if="showErrors"/>
    </slot>
  </form>
</template>
<script>
import { useVueform, Vueform } from '@vueform/vueform';

export default {
  mixins: [Vueform],
  setup: useVueform,
}
</script>
```
After registering `CustomVueform` globally, you can use is as:
```vue
<template>
  <CustomVueform endpoint="https://xxxxx.xx/xxxxx" method="post" @response="handleResponse">  
   <TextElement name="name" label="Name" rules="required|max:80" />
   <ButtonElement name="submit" submits> Submit </ButtonElement>
  </CustomVueform>
</template>
```

**Way 3:** You can create a [custom template](https://vueform.com/docs/styles-and-layout#replacing-templates) for Vueform and override the default globally:
```vue
<!-- CustomVueformTemplate.vue -->
<template>
  <form :class="classes.form" @submit.prevent="submit">
    <slot name="empty">
      <FormElements><slot/></FormElements>
      <component is="FormMessages" v-if="showMessages"/>
      <component is="FormErrors" v-if="showErrors"/>
    </slot>
  </form>
</template>
<script>
  export default {
    data() {
      return {
        merge: true,
        defaultClasses: {
          form: '',
        }
      }
    }
  }
</script>
```

Then [registering](https://vueform.com/reference/configuration#templates) it globally:

```js
// vueform.config.js

import CustomVueformTemplate from '/CustomVueformTemplate.vue'

export default {
  templates: {
    Vueform: CustomVueformTemplate,
  }
  // ...
}
```

**Way 4:** You can register the same template as an alternative `view` if you don't want to override the default:
```js
// vueform.config.js

import CustomVueformTemplate from '/CustomVueformTemplate.vue'

export default {
  templates: {
    Vueform_bottomMsg: CustomVueformTemplate,
  }
  // ...
}
```

Then use it whenever needed:
```vue
<template>
  <Vueform ... view="bottomMsg">
    <TextElement name="name" label="Name" rules="required|max:80" />
    <ButtonElement name="submit" submits> Submit </ButtonElement>
  </Vueform>
</template>
```

**Reply by nvijayap** [reply]
Thanks for the very detailed response, I appreciate. Tried something very simple based on your response and ran into this disconnect - unable to pick up the specific fields of the response using the dot notation for some reason ...
<img width="323" alt="image" src="https://github.com/vueform/vueform/assets/881245/6348592c-05e0-4aea-8118-041d6a9ad6ca">

**Reply by nvijayap** [reply]
The JSON in the upper part of above `image` shows the `message` and `status` clearly. However, the lower part of above image shows the `message` as `undefined`. The `image` was picked up from the browser's Web Developer Tools (`Firefox's Web Developer Tools`).

**Reply by nvijayap** [reply]
BTW, `status` (`int`) shows up fine ...
<img width="308" alt="image" src="https://github.com/vueform/vueform/assets/881245/8fe60310-c77e-4386-8cc1-06b042a1037d">

**Reply by adamberecz** [reply]
You need `response.data.message`.

The `response` is an [Axios response object](https://axios-http.com/docs/res_schema) (see our [reference](https://vueform.com/reference/vueform#event-response)), which also happens to have a `status` prop. Everything you send in the response body can be accessed via `response.data`.

**Reply by nvijayap** [reply]
Thank you, that worked, I appreciate

**Comment by delano** [comment]
> ```vue
> <script setup>
> const handleResponse = (response, form$) => {
>   // response is {}, form$ is undefined
> }
> </script>
> 
> <template>
>   <Vueform endpoint="https://xxxxx.xx/xxxxx" method="post" @response="handleResponse">
>     <TextElement name="name" label="Name" rules="required|max:80" />
>     <ButtonElement name="submit" submits> Submit </ButtonElement>
>   </Vueform>
> </template>
> ```
>

@adamberecz I'm trying out [this example above](https://github.com/vueform/vueform/discussions/74#discussioncomment-7684021) using the `response` event handler. The `handleResponse` function is only receiving one argument and it's just a plain Object (no reference to the response or form).  Checking the `success` and `error` events, they both receive an `AxiosError` object rather than an `AxiosResponse`. 

I thought it might have to do with how I had the axios instance configured so I'm using the [`vueform.config.ts` from the installation docs ](https://vueform.com/docs/installation) verbatim. I'm running v1.6.3. 

Does this sounds familiar at all? I'm assuming it's something I'm doing. Happy to dig into it further if need be.

**Comment by gevezex** [comment]
@adamberecz Thnx for these examples. The only thing is when the backend is not responding (when submitting to a backend), there is no error as there is no return code. What is the way to also catch that kind of an error?
